### PR TITLE
【PaddlePaddle Hackathon 4】add einsum fp16 test and add Heaviside bf16 support

### PR DESCRIPTION
**Pull request by AndPuQing**
### PR types
New features

### PR changes
OPs

### Describe
- add einsumop fp16 test
- add heaviside bf16 and test

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 1]
```diff
@@ -2164,7 +2164,7 @@ def _assert_is_close(
                 else:
                     abs_a = 1 if abs_a < 1e-3 else abs_a
 
-            if self.dtype == np.bool:
+            if self.dtype == np.bool_:
```
Comment (201716010711) [inline python/paddle/fluid/tests/unittests/op_test.py]: 不要改op_test文件

File: python/paddle/fluid/tests/unittests/test_einsum_v2.py [hunk 2]
```diff
@@ -557,6 +557,24 @@ def test_shape(self):
...
+        A = paddle.to_tensor(np.array([1.0, 2.0])).astype(paddle.float16)
+        A = A.cuda()
+        B = paddle.to_tensor(np.array([2.0, 3.0])).astype(paddle.float16)
+        B = B.cuda()
+        C = paddle.einsum('i,i->', A, B)
+        D = paddle.to_tensor(8.0).astype(paddle.float16)
+        self.assertEqual(C.item(), D.item())
+
+
 class TestComplex(unittest.TestCase):
```
Comment (201716010711) [inline python/paddle/fluid/tests/unittests/test_einsum_v2.py:587]: 添加FP16单测可以参考pr：https://github.com/PaddlePaddle/Paddle/pull/50723

File: python/paddle/fluid/tests/unittests/test_elementwise_heaviside_op.py [hunk 3]
```diff
@@ -185,6 +186,48 @@ def test_check_grad(self):
         )
 
 
+@unittest.skipIf(
+    not paddle.is_compiled_with_cuda(), "core is not compiled with CUDA"
+)
+class TestHeavisideAPI_bfloat16(OpTest):
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_elementwise_heaviside_op.py:192]: TestHeavisideAPI_bfloat16改为TestHeavisideBF16OP

File: python/paddle/fluid/tests/unittests/test_elementwise_heaviside_op.py [hunk 4]
```diff
@@ -185,6 +186,48 @@ def test_check_grad(self):
...
+        place = core.CUDAPlace(0)
+        self.check_output_with_place(place)
+
+    def test_check_grad(self):
+        place = core.CUDAPlace(0)
+        self.check_grad_with_place(
+            place,
+            ['X', 'Y'],
+            'Out',
+            user_defined_grads=Heaviside_grad(
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_elementwise_heaviside_op.py:225]: 先用不带有user_defined_grads参数的调用尝试

File: python/paddle/fluid/tests/unittests/test_einsum_v2.py [hunk 5]
```diff
@@ -159,63 +159,75 @@ def _get_place(self, force_to_use_cpu=False):
                 return core.CUDAPlace(0)
             return core.CPUPlace()
 
-    def check_output_equal(self, actual, expect, rtol=1.0e-5, atol=1.0e-8):
+    def check_output_equal(
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_einsum_v2.py:162]: 最好添加OpTest框架的单测

**Pull request closed without merging**